### ISSUE TITLE
ClassNotFoundException: com.google.common.base.Joiner when activated by pom.xml

### ISSUE THREAD

**Issue by seanf** [takari/takari-smart-builder#6]
Using configuration like this in pom.xml:
```xml
  <build>
    <extensions>
      <!-- Enable Takari's Concurrent Safe Local Repository -->
      <extension>
        <groupId>io.takari.aether</groupId>
        <artifactId>takari-local-repository</artifactId>
        <version>0.11.2</version>
      </extension>
      <extension>
        <groupId>io.takari</groupId>
        <artifactId>takari-filemanager</artifactId>
        <version>0.8.3</version>
      </extension>
      <!-- Enable Takari's Smart Builder -->
      <extension>
        <groupId>io.takari.maven</groupId>
        <artifactId>takari-smart-builder</artifactId>
        <version>0.4.1</version>
      </extension>
    </extensions>
  </build>
```

I get a ClassNotFoundException for a Guava class:

	$ mvn package --builder smart -T1.0C
	12:23:57 [INFO] Scanning for projects...
	12:23:58 [INFO]
	12:23:58 [INFO] Using the SmartBuilder implementation with a thread count of 8
	12:23:58 [INFO] ------------------------------------------------------------------------
	12:23:58 [INFO] BUILD SUCCESS
	12:23:58 [INFO] ------------------------------------------------------------------------
	12:23:58 [INFO] Total time: 0.187 s (Wall Clock)
	12:23:58 [INFO] Finished at: 2016-11-01T12:23:58+10:00
	12:23:58 [INFO] Final Memory: 8M/298M
	12:23:58 [INFO] ------------------------------------------------------------------------
	---------------------------------------------------
	constituent[0]: file:/home/sflaniga/apps/apache-maven/lib/ext/takari-smart-builder-0.4.1.jar
	constituent[1]: file:/home/sflaniga/apps/apache-maven/lib/ext/maven-notifier-1.9.1-shaded.jar
	constituent[2]: file:/home/sflaniga/apps/apache-maven/lib/ext/takari-filemanager-0.8.3.jar
	constituent[3]: file:/home/sflaniga/apps/apache-maven/lib/ext/takari-local-repository-0.11.2.jar
	constituent[4]: file:/home/sflaniga/apps/apache-maven/lib/javax.inject-1.jar
	constituent[5]: file:/home/sflaniga/apps/apache-maven/lib/maven-artifact-3.3.9.jar
	constituent[6]: file:/home/sflaniga/apps/apache-maven/lib/aether-spi-1.0.2.v20150114.jar
	constituent[7]: file:/home/sflaniga/apps/apache-maven/lib/guice-4.0-no_aop.jar
	constituent[8]: file:/home/sflaniga/apps/apache-maven/lib/plexus-utils-3.0.22.jar
	constituent[9]: file:/home/sflaniga/apps/apache-maven/lib/jsr250-api-1.0.jar
	constituent[10]: file:/home/sflaniga/apps/apache-maven/lib/slf4j-api-1.7.5.jar
	constituent[11]: file:/home/sflaniga/apps/apache-maven/lib/plexus-sec-dispatcher-1.3.jar
	constituent[12]: file:/home/sflaniga/apps/apache-maven/lib/maven-compat-3.3.9.jar
	constituent[13]: file:/home/sflaniga/apps/apache-maven/lib/maven-aether-provider-3.3.9.jar
	constituent[14]: file:/home/sflaniga/apps/apache-maven/lib/maven-settings-builder-3.3.9.jar
	constituent[15]: file:/home/sflaniga/apps/apache-maven/lib/wagon-http-2.10-shaded.jar
	constituent[16]: file:/home/sflaniga/apps/apache-maven/lib/maven-model-builder-3.3.9.jar
	constituent[17]: file:/home/sflaniga/apps/apache-maven/lib/cdi-api-1.0.jar
	constituent[18]: file:/home/sflaniga/apps/apache-maven/lib/org.eclipse.sisu.inject-0.3.2.jar
	constituent[19]: file:/home/sflaniga/apps/apache-maven/lib/aether-api-1.0.2.v20150114.jar
	constituent[20]: file:/home/sflaniga/apps/apache-maven/lib/plexus-cipher-1.7.jar
	constituent[21]: file:/home/sflaniga/apps/apache-maven/lib/wagon-file-2.10.jar
	constituent[22]: file:/home/sflaniga/apps/apache-maven/lib/maven-builder-support-3.3.9.jar
	constituent[23]: file:/home/sflaniga/apps/apache-maven/lib/commons-io-2.2.jar
	constituent[24]: file:/home/sflaniga/apps/apache-maven/lib/aether-transport-wagon-1.0.2.v20150114.jar
	constituent[25]: file:/home/sflaniga/apps/apache-maven/lib/wagon-provider-api-2.10.jar
	constituent[26]: file:/home/sflaniga/apps/apache-maven/lib/aether-util-1.0.2.v20150114.jar
	constituent[27]: file:/home/sflaniga/apps/apache-maven/lib/aether-impl-1.0.2.v20150114.jar
	constituent[28]: file:/home/sflaniga/apps/apache-maven/lib/org.eclipse.sisu.plexus-0.3.2.jar
	constituent[29]: file:/home/sflaniga/apps/apache-maven/lib/commons-cli-1.2.jar
	constituent[30]: file:/home/sflaniga/apps/apache-maven/lib/maven-embedder-3.3.9.jar
	constituent[31]: file:/home/sflaniga/apps/apache-maven/lib/plexus-interpolation-1.21.jar
	constituent[32]: file:/home/sflaniga/apps/apache-maven/lib/jsoup-1.7.2.jar
	constituent[33]: file:/home/sflaniga/apps/apache-maven/lib/maven-plugin-api-3.3.9.jar
	constituent[34]: file:/home/sflaniga/apps/apache-maven/lib/aopalliance-1.0.jar
	constituent[35]: file:/home/sflaniga/apps/apache-maven/lib/maven-core-3.3.9.jar
	constituent[36]: file:/home/sflaniga/apps/apache-maven/lib/plexus-component-annotations-1.6.jar
	constituent[37]: file:/home/sflaniga/apps/apache-maven/lib/maven-settings-3.3.9.jar
	constituent[38]: file:/home/sflaniga/apps/apache-maven/lib/slf4j-simple-1.7.5.jar
	constituent[39]: file:/home/sflaniga/apps/apache-maven/lib/aether-connector-basic-1.0.2.v20150114.jar
	constituent[40]: file:/home/sflaniga/apps/apache-maven/lib/guava-18.0.jar
	constituent[41]: file:/home/sflaniga/apps/apache-maven/lib/wagon-http-shared-2.10.jar
	constituent[42]: file:/home/sflaniga/apps/apache-maven/lib/maven-model-3.3.9.jar
	constituent[43]: file:/home/sflaniga/apps/apache-maven/lib/maven-repository-metadata-3.3.9.jar
	constituent[44]: file:/home/sflaniga/apps/apache-maven/lib/commons-lang-2.6.jar
	constituent[45]: file:/home/sflaniga/apps/apache-maven/lib/commons-lang3-3.4.jar
	constituent[46]: file:/home/sflaniga/apps/apache-maven/conf/logging/
	---------------------------------------------------
	Exception in thread "main" java.lang.NoClassDefFoundError: com/google/common/base/Joiner
	        at io.takari.maven.builder.smart.SmartBuilderImpl.build(SmartBuilderImpl.java:152)
	        at io.takari.maven.builder.smart.SmartBuilder.build(SmartBuilder.java:31)
	        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	        at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	        at java.lang.reflect.Method.invoke(Method.java:498)
	        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	Caused by: java.lang.ClassNotFoundException: com.google.common.base.Joiner
	        at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	        at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	        at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	        at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	        ... 17 more

The error message is similar to the one in https://github.com/jcgay/maven-color/issues/10, but I haven't tried the workaround: https://github.com/jcgay/maven-color/issues/10#issuecomment-111717689

I do not see this problem when running Smart Builder from `$M2/lib/ext/` or `.mvn/extensions.xml`, only when activated by the pom. I only just realised that I probably should have been using `.mvn/extensions.xml`, not `pom.xml`, so this is probably not expected to work, but even so, it would be better to have a meaningful error message, not ClassNotFoundException.

Sample project: [my-app.zip](https://github.com/takari/takari-smart-builder/files/563211/my-app.zip)

**Comment by ifedorenko** [comment]
Indeed, SmartBuidler cannot be enabled in pom.xml file, you need to use `.mvn/extensions.xml` or modify your maven installation. 

Also,  Takari local repository is not thread safe, you really don't want to use it with SmartBuilder. Use standard Maven local repo implementation, it works fine with single-process multithreaded build (it may work for local repository shared among multiple builds too, I just haven't tested multi-build usecase much).

**Comment by seanf** [comment]
I thought the whole point of Takari local repository was to be ["safe for concurrent access from multiple threads and Maven invocations"](https://github.com/takari/takari-local-repository) (also documented [here](http://takari.io/book/30-team-maven.html)). @ifedorenko Are you saying that it doesn't work as documented?

**Comment by ifedorenko** [comment]
takari local repository does not correctly work in multithreaded build (see a test I provided in https://github.com/takari/takari-local-repository/issues/5), and other users report it does not work with local repository shared among multiple build processes (https://github.com/takari/takari-local-repository/issues/4)

**Comment by seanf** [comment]
Thanks for the warning @ifedorenko.

Incidentally, I tried your test project https://github.com/ifedorenko/takari-local-repository-race on Windows 10, reproduced the artifact resolution error from https://github.com/takari/takari-local-repository/issues/5. 

Then I disabled takari-local-repository and instead of fixing the problem, I got the "....jar.part.lock (Access is denied)" error from https://github.com/takari/takari-local-repository/issues/4. So I don't think the standard repository code can be trusted (for multi-threaded access) either, unfortunately.
